binary_heap.py

```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+
+
+class BinaryHeap(object):
+    """Implements a binary max-heap data structure"""
+
+    def __init__(self, an_iter=None):
+        '''Initializes the heap,
+        optionally populates it with the given iterable
+        '''
+
+        self._list = []
+
+        if an_iter:
+            for val in an_iter:
+                self.push(val)
+
+    def push(self, val):
+        '''Pushes the value into the heap,
+         adjusting the ordering to maintain the heap
+         '''
+
+        self._list.append(val)
+        child_pos = len(self._list) - 1
+
+        while child_pos and (
+                self._list[child_pos] > self._list[(child_pos - 1) // 2]):
+
+            # swap parent and child
+            self._switch(child_pos)
+            child_pos = (child_pos - 1) // 2
+
+    def pop(self):
+        '''Pops the top value off the heap and returns it,
+         adjusting the ordering accordingly
+         '''
+
+        try:
+            self._list[0], self._list[-1] = self._list[-1], self._list[0]
+            top = self._list.pop()
+
+            parent_pos = 0
+            children_pos = self._find_children(parent_pos)
+            while children_pos != []:
+                larger_child = None
+                larger_child_index = None
+                if len(children_pos) > 1:
+                    if self._list[children_pos[0]] >= self._list[
+                            children_pos[1]]:
+                        larger_child = self._list[children_pos[0]]
+                        larger_child_index = children_pos[0]
+                    else:
+                        larger_child = self._list[children_pos[1]]
+                        larger_child_index = children_pos[1]
+                else:
+                    larger_child = self._list[children_pos[0]]
+                    larger_child_index = children_pos[0]
+
+                if self._list[parent_pos] < larger_child:
+                    self._switch(larger_child_index)
+                else:
+                    break
+                parent_pos = larger_child_index
+                children_pos = self._find_children(parent_pos)
```

Comment:
This is a pretty big blob of code with a lot of nested if/else stuff.  Breaking it out into one or more helper functions would reduce visual complexity and allow for testing it outside the scope of he `pop` method.